### ISSUE TITLE
[ICE] rustc panics with 'Trying to convert unsized value to lval'

### ISSUE THREAD

**Issue by alexchandel** [rust-lang/rust#22829]
rustc panics with the following backtrace when building [this fork of rlibc](https://github.com/alexchandel/rlibc). I can't even tell which line of my code is causing it, but it occurs after most of the warnings are printed:

```sh
...
src/posix/fcntl.rs:6:62: 6:66 warning: unused variable: `mode`, #[warn(unused_variables)] on by default
src/posix/fcntl.rs:6 pub unsafe extern fn open(path: *const char_t, flags: int_t, mode: mode_t) -> int_t {
                                                                                  ^~~~
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Trying to convert unsized value to lval', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/librustc_trans/trans/datum.rs:495

stack backtrace:
   1:        0x112612be3 - sys::backtrace::write::h918207864656307b0nC
   2:        0x112640be5 - panicking::on_panic::h2c4b1456225ae37fnPL
   3:        0x11256a418 - rt::unwind::begin_unwind_inner::h6c7f1344c68dd4a7AwL
   4:        0x10ea6dc45 - rt::unwind::begin_unwind::h4082345751407454387
   5:        0x10eaec56f - trans::datum::Datum<'tcx, Expr>::to_lvalue_datum::h4809841e9861f286hYf
   6:        0x10eb20dd5 - trans::expr::trans_index::h56dd6242817519cf6Ei
   7:        0x10eb0d9c9 - trans::expr::trans_unadjusted::h8c78f89bb783464csqi
   8:        0x10eac21d0 - trans::expr::trans::h25c26cd42d9241e1HMh
   9:        0x10eb237e3 - trans::expr::trans_addr_of::h5be1b2f119f51174rNj
  10:        0x10eb0dbd7 - trans::expr::trans_unadjusted::h8c78f89bb783464csqi
  11:        0x10eac21d0 - trans::expr::trans::h25c26cd42d9241e1HMh
  12:        0x10eafdba5 - trans::callee::trans_args::hc52bfad7cdf49353xjh
  13:        0x10eb048fc - trans::callee::trans_call_inner::h5306915459861161960
  14:        0x10eb0fdaa - trans::expr::trans_rvalue_dps_unadjusted::hb1406d54d45518409Wi
  15:        0x10eb0d642 - trans::expr::trans_unadjusted::h8c78f89bb783464csqi
  16:        0x10eac21d0 - trans::expr::trans::h25c26cd42d9241e1HMh
  17:        0x10eb22c0c - trans::expr::trans_unary::h509c840b305ef674LFj
  18:        0x10eb0dbc0 - trans::expr::trans_unadjusted::h8c78f89bb783464csqi
  19:        0x10eac0a0b - trans::expr::trans_into::h0f31dfb1ff207cd8pGh
  20:        0x10ebb5fd7 - trans::_match::trans_match_inner::h56e2641ee60bee0daMx
  21:        0x10eb0f1b3 - trans::expr::trans_rvalue_dps_unadjusted::hb1406d54d45518409Wi
  22:        0x10eac09eb - trans::expr::trans_into::h0f31dfb1ff207cd8pGh
  23:        0x10eac17fa - trans::controlflow::trans_block::hb675814db44e1ef35ee
  24:        0x10eb982bf - trans::base::trans_closure::h1fc9dae5b42c150ephu
  25:        0x10eaac086 - trans::base::trans_fn::h1baa1a79c6642379Gsu
  26:        0x10eaa6fc9 - trans::base::trans_item::h99ee5fcee3ea0222zRu
  27:        0x10eb99a89 - trans::base::trans_mod::h0f4f7f5d04ce3f65yXu
  28:        0x10eaa70dd - trans::base::trans_item::h99ee5fcee3ea0222zRu
  29:        0x10eb9ee0c - trans::base::trans_crate::h89e3520c5dfd05b21Ov
  30:        0x10e92d058 - driver::phase_4_translate_to_llvm::h5a7c1f8c7ec04632qPa
  31:        0x10e907f72 - driver::compile_input::h32723f720e0c3c95Gba
  32:        0x10e9e08a7 - run_compiler::hfdf8469da778a8d4Zbc
  33:        0x10e9ddd21 - thunk::F.Invoke<A, R>::invoke::h56535309890628509
  34:        0x10e9dc980 - rt::unwind::try::try_fn::h6830563953582217021
  35:        0x1126ba7e9 - rust_try_inner
  36:        0x1126ba7d6 - rust_try
  37:        0x10e9dd0e5 - thunk::F.Invoke<A, R>::invoke::h12775416177659868106
  38:        0x112629ac3 - sys::thread::thread_start::hccc0587963a6c50bU4G
  39:     0x7fff8cc79268 - _pthread_body
  40:     0x7fff8cc791e5 - _pthread_body

make: *** [target/x86_64-apple-darwin/c.o] Error 101
```

**Comment by SimonSapin** [comment]
This is possibly a duplicate of #21847, which seems fixed in rustc 1.0.0-nightly (4db0b32 2015-02-25) (built 2015-02-26). Can you still reproduce on a recent nigthly?

**Comment by alexchandel** [comment]
Seems to be fixed.
